autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java

```diff
@@ -0,0 +1,98 @@
+                    grpcUmsClient.assignResourceRole(machineUserCrn, environmentCrn,
+                            roleCrnGenerator.getBuiltInEnvironmentUserResourceRoleCrn(), MDCUtils.getRequestId());
+                    LOGGER.info("Assigned resourcerole '{}' for  machineUserCrn '{}' for environment '{}'",
+                            roleCrnGenerator.getBuiltInEnvironmentUserResourceRoleCrn(), machineUserCrn, environmentCrn);
+                }
+                syncEnvironment(cluster.getClusterPertain().getTenant(), machineUserCrn, environmentCrn, Optional.empty());
+                clusterService.setMachineUserCrn(cluster.getId(), machineUserCrn);
+            }
+        } catch (Exception ex) {
+            LOGGER.warn("Error initializing machineUserCrn for cluster '{}' yarn polling", cluster.getStackCrn(), ex);
+        }
+    }
+
+    @Retryable(value = Exception.class, maxAttempts = 5, backoff = @Backoff(delay = 10000))
+    public void deleteMachineUserForEnvironment(String accountId, String machineUserCrn, String environmentCrn) {
+        if (environmentCrn != null && machineUserCrn != null) {
+            MachineUser machineUser = getOrCreateAutoscaleMachineUser(environmentCrn, accountId);
+            grpcUmsClient.deleteMachineUser(machineUser.getCrn(), ThreadBasedUserCrnProvider.INTERNAL_ACTOR_CRN, accountId, MDCUtils.getRequestId());
+            syncEnvironment(accountId, machineUserCrn, environmentCrn, Optional.of(machineUser.getWorkloadUsername()));
+            LOGGER.info("Deleted MachineUser for machineUserCrn '{}', environment '{}'", machineUserCrn, environmentCrn);
+        }
+    }
+
+    private MachineUser getOrCreateAutoscaleMachineUser(String environmentCrn, String accountId) {
+        //Idempotent api retrieves machine user or creates if missing.
+        String autoscaleMachineUserName = String.format(AUTOSCALE_MACHINE_USER_NAME_PATTERN, Crn.fromString(environmentCrn).getResource());
+        MachineUser machineUser = grpcUmsClient.getOrCreateMachineUserWithoutAccessKey(autoscaleMachineUserName, accountId, MDCUtils.getRequestId());
+        LOGGER.info("Retrieved machineUser '{}' for machineUserName '{}' ", machineUser, autoscaleMachineUserName);
+        return machineUser;
+    }
+
+    private void syncEnvironment(String accountId, String machineUserCrn, String environmentCrn, Optional<String> deletedWorkloadUserName) {
+        ThreadBasedUserCrnProvider.doAsInternalActor(() -> {
+            SynchronizeAllUsersRequest request = new SynchronizeAllUsersRequest();
+            request.setAccountId(accountId);
+            request.setEnvironments(Set.of(environmentCrn));
+            request.setMachineUsers(Set.of(machineUserCrn));
+            if (deletedWorkloadUserName.isPresent()) {
+                request.setDeletedWorkloadUsers(Set.of(deletedWorkloadUserName.get()));
+            }
+            userV1Endpoint.synchronizeAllUsers(request);
```

Comment:
please check if this manual sync is still needed

Reply:
When the patch was initially written - automatic sync was not available, and https://jira.cloudera.com/browse/CB-14554 was created as a follow up to remove this sync.

If the automatic user-sync is in place now - we should remove this. Think it'll need some testing though.
Specifically: 
1. Does the automatic sync actually happen.
2. How long does it end up taking. (Don't want to run into a situation where a user is created, we try using it to query YARN metrics, and end up failing)

Reply:
the API calls to FMS will not wait until syncs are finished, just will initialize them afaik
so you probably need to check with different call if sync is finished

Reply:
> Automatic sync is implemented for tenants with the WORKLOAD_IAM_SYNC entitlement enabled. For tenants without this entitlement, a manual sync is required.

So it sounds like the sync is still required, and the linked jira stays open until the sync is enabled by default.

> Unsure about the exact duration of the sync, however, if there are many powerUsers / users granted with many roles, then the sync might take longer. In the YarnMetricsClient class, if machineUserCrn is not intialised (creation ended up failing), then it falls back to the orginal userCrn.

The YarnMetricsClient class checks for the existence of the user in the DB, which in itself does not guarantee an IPA sync. Linking this with a comment later on - can we determine reason for failure, or even in the absence of this - fallback to the cluster creator, and try again later - given user-sync adequate time to perform the sync. The state for the failure can be stored in the DB.

Curious about how the initial user-sync happens for an environment. This would all be so much simpler if a single env level user could be created before this step (at least for new environemtn). @lacikaaa - any idea on the first user-sync flow? Maybe that could be used for reference as well.

Reply:
@sidseth Initial usersync happens right after the environment creation, usually happens parallel with datalake creation. At that point there is no datahub in the given environment.

An idea can be to prepare couple of machine users with a naming convention during initial sync and use them later for datahubs. But it sound only a workaround since we cannot guess how many datahubs will be created in an environment.

Reply:
The user being created here is not linked to the DataHub. I believe it is linked to the environment itself. So maybe we move the user-creation to the time when the env is created, prior to the first user-sync. (The current patch will likely still be required for envs where this user has not been created).
Just a thought ... @horadla23  -  do you have a code pointer for this? To see if we can create this user up-front?

Reply:
Ohh yes right, I forgot this will be a machine user on environment level. then this class can be suitable for this purpose:https://github.com/hortonworks/cloudbreak/blob/6504996cb652093b22202d3c2ada54753929b837/environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/EnvironmentInitHandler.java#L107
here we are assigning EnvironmentAdmin to the creator. After that, machine user can be created and EnvironmentUser can be assigned and no additional user sync will needed.